### PR TITLE
Feature/todo list

### PR DESCRIPTION
**Pull request by liaarbel**
I created a tasks list for every day in `dayview` by clicking on the button `create task` you can add task to the daily tasks list.
by clicking on the button `todo list` you can see all your daily tasks sorted by time and marked by important or not.
after you finish a task you can mark the task `Done`.

### CODE REVIEW

File: app/database/models.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 from sqlalchemy import (
     Boolean, Column, DateTime, DDL, event, Float, ForeignKey, Index, Integer,
-    JSON, String, Time, UniqueConstraint)
+    JSON, String, Time, Date, UniqueConstraint)
```
Comment (yammesicka) [inline app/database/models.py]: Please sort by the ABC

File: app/internal/todo_list.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        owner_id=owner_id,
+        is_important=is_important,
+        is_done=False
+    )
+    return task
+
+
+def sort_by_time(tasks: List[Task]) -> List[Task]:
+    """Sorts the tasks by the start of the task."""
+
```
Comment (yammesicka) [inline app/internal/todo_list.py]: Avoid adding new line after docstrings in functions

File: app/internal/todo_list.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        date=datestr,
+        time=timestr,
+        owner_id=owner_id,
+        is_important=is_important,
+        is_done=False
+    )
+    return task
+
+
+def sort_by_time(tasks: List[Task]) -> List[Task]:
```
Comment (yammesicka) [inline app/internal/todo_list.py]: Prefer sorting in the database, it's probably faster

File: app/internal/todo_list.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    """Sorts the tasks by the start of the task."""
+
+    temp = tasks.copy()
+    return sorted(temp, key=attrgetter('time'))
+
+
+def by_id(db, task_id):
```
Comment (yammesicka) [inline app/internal/todo_list.py]: Please add type annotations to your functions

File: app/routers/dayview.py [hunk 5]
```diff
@@ -5,10 +5,11 @@
 from fastapi import APIRouter, Depends, HTTPException, Request
 from fastapi.templating import Jinja2Templates
 
-from app.database.models import Event, User
+from app.database.models import Event, User, Task
```
Comment (yammesicka) [inline app/routers/dayview.py]: ABC

File: app/routers/todo_list.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+from fastapi import Depends, APIRouter, Form
+from fastapi.encoders import jsonable_encoder
+from requests import Session
+from sqlalchemy.exc import SQLAlchemyError
+from starlette import status
+from starlette.responses import RedirectResponse, JSONResponse
+
+from app.config import templates
+from app.dependencies import get_db
+from app.database.models import User
```
Comment (yammesicka) [inline app/routers/todo_list.py]: abc

File: app/routers/todo_list.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    task.title = title
+    task.description = description
+    task.date = datetime.strptime(datestr, '%Y-%m-%d').date()
+    task.time = datetime.strptime(timestr, '%H:%M:%S').time()
+    task.is_important = is_important
+    session.commit()
+    return RedirectResponse(f"/day/{datestr}", status_code=303)
+
+
+@router.post("/setDone/{task_id}")
```
Comment (yammesicka) [inline app/routers/todo_list.py]: /done/{task_id} (the "set" is implicit because it is a `POST` request)

File: app/routers/todo_list.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+async def set_task_done(task_id: int, session=Depends(get_db)):
+    task = by_id(session, task_id)
+    task.is_done = True
+    session.commit()
+    return RedirectResponse(f"/day/{task.date.strftime('%Y-%m-%d')}",
+                            status_code=303)
+
+
+@router.post("/setUndone/{task_id}")
```
Comment (yammesicka) [inline app/routers/todo_list.py]: /undone/

File: app/static/dayview.css [hunk 9]
```diff
@@ -106,4 +106,91 @@ html {
...
   border-radius: 50px;
   box-shadow: 1px 1px 2px #999;
+}
+
+.todo-style {
+	background: #F7F7F7;
+}
+
+#table {
+    max-width: 500px;
```
Comment (yammesicka) [inline app/static/dayview.css]: % or em

File: app/static/dayview.css [hunk 10]
```diff
@@ -106,4 +106,91 @@ html {
...
+}
+
+.open-button {
+    background-color: var(--primary);
+    color: #fff;
+    margin-top: 12px;
+    font-size: 20px;
+    padding: 14px;
+    border-radius: 20px;
+    transition-duration: 0.4s;
```
Comment (yammesicka) [inline app/static/dayview.css]: Change all the values (except maybe border radius) to em

File: app/templates/dayview.html [hunk 11]
```diff
@@ -1,64 +1,267 @@
...
+    <meta name="viewport" content="width=device-width, initial-scale=1">
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.0-beta1/dist/css/bootstrap.min.css"
+          rel="stylesheet"
+          integrity="sha384-giJF6kkoqNQ00vy+HMDP7azOuL0xtbfIcaT9wjKHr8RbDVddVHyTfAAsrekwKmP1"
+          crossorigin="anonymous">
+    <link rel="stylesheet" href="{{ url_for('static', path='/dayview.css') }}">
+    <link href="https://fonts.googleapis.com/css2?family=Assistant&family=Open+Sans&family=Roboto&family=Roboto+Condensed&display=swap"
+          rel="stylesheet">
+    <title>dayview</title>
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
```
Comment (yammesicka) [inline app/templates/dayview.html]: Remove the jQuery

File: app/templates/dayview.html [hunk 12]
```diff
@@ -1,64 +1,267 @@
...
+    <script type="text/javascript">
```
Comment (yammesicka) [inline app/templates/dayview.html]: Put in a different JS file

**Review by yammesicka — COMMENTED** [review]
That's a good start! I have added few insights :)

File: app/database/models.py (hunk 1 above)
Comment (liaarbel) [reply app/database/models.py]: fixed

File: app/internal/todo_list.py (hunk 2 above)
Comment (liaarbel) [reply app/internal/todo_list.py]: fixed

File: app/internal/todo_list.py (hunk 3 above)
Comment (liaarbel) [reply app/internal/todo_list.py]: But the most logical thing to do is to sort your daily task by the time they happen

File: app/internal/todo_list.py (hunk 4 above)
Comment (liaarbel) [reply app/internal/todo_list.py]: fixed

File: app/routers/dayview.py (hunk 5 above)
Comment (liaarbel) [reply app/routers/dayview.py]: fixed

File: app/routers/todo_list.py (hunk 6 above)
Comment (liaarbel) [reply app/routers/todo_list.py]: fixed

File: app/routers/todo_list.py (hunk 7 above)
Comment (liaarbel) [reply app/routers/todo_list.py]: fixed

File: app/routers/todo_list.py (hunk 8 above)
Comment (liaarbel) [reply app/routers/todo_list.py]: fixed

File: app/static/dayview.css (hunk 10 above)
Comment (liaarbel) [reply app/static/dayview.css]: fixed

File: app/static/dayview.css (hunk 9 above)
Comment (liaarbel) [reply app/static/dayview.css]: fixed

File: app/templates/dayview.html (hunk 11 above)
Comment (liaarbel) [reply app/templates/dayview.html]: But I'm using this lib

Comment (yammesicka) [reply app/templates/dayview.html]: Where? Maybe I can help you remove it.
We don't use jQuery 'cause the size of it cost a lot (large size, so it's greatly slows the website) and can almost always be replaced with something else pretty easily

File: tests/test_todo_list.py [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+
+@pytest.fixture
+def test1(session):
+    user_example = create_model(
+        session, User, username="test_username", email="testtodo@gmail.com",
+        password="123456", full_name="My test", language_id=1
+    )
+    yield user_example
+    delete_instance(session, user_example)
```
Comment (Gonzom) [inline tests/test_todo_list.py]: See if you can use the existing user fixture instead of creating a new one.

File: tests/test_todo_list.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+                                           "timestr": task.time.strftime(
+                                               '%H:%M'),
+                                           "is_important": task.is_important,
+                                           "session": session})
+    assert response.status_code == 303
+    task_ex = by_id(session, task.id)
+    assert task == task_ex
+
+
+def test_if_task_edited(home_test_client, session, task2: Task):
```
Comment (Gonzom) [inline tests/test_todo_list.py]: Can you use task(1) here instead? I don't see anything special about task2 here.

File: app/main.py [hunk 15]
```diff
@@ -62,6 +62,7 @@ def create_tables(engine, psql_environment):
     search.router,
     telegram.router,
     whatsapp.router,
+    todo_list.router,
```
Comment (ivarshav) [inline app/main.py:137]: please keep alphabetic order :)

File: app/routers/todo_list.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+    return RedirectResponse(f"/day/{datestr}", status_code=303)
+
+
+@router.post("/setDone/{task_id}")
+async def set_task_done(task_id: int, session=Depends(get_db)):
+    task = by_id(session, task_id)
+    task.is_done = True
+    session.commit()
+    return RedirectResponse(f"/day/{task.date.strftime('%Y-%m-%d')}",
+                            status_code=303)
```
Comment (ivarshav) [inline app/routers/todo_list.py]: use const for status

File: app/internal/todo_list.py (hunk 3 above)
Comment (yammesicka) [reply app/internal/todo_list.py]: You're right - I meant to order them when you querying the database using `sort by`

File: tests/test_todo_list.py (hunk 14 above)
Comment (liaarbel) [reply tests/test_todo_list.py]: But I change the data to test if it edited.

File: app/routers/todo_list.py (hunk 16 above)
Comment (liaarbel) [reply app/routers/todo_list.py:77]: What do you mean?

File: app/templates/partials/calendar/calendar_base.html [hunk 17]
```diff
@@ -20,5 +20,7 @@
 <script src="{{ url_for('static', path='/js/grid_scripts.js') }}"></script>
 <script src="{{ url_for('static', path='/js/grid_navigation.js') }}"></script>
 <script src="https://unpkg.com/sweetalert/dist/sweetalert.min.js"></script>
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
```
Comment (yammesicka) [inline app/templates/partials/calendar/calendar_base.html]: Remove

File: app/main.py (hunk 15 above)
Comment (liaarbel) [reply app/main.py:137]: fixed

File: app/routers/todo_list.py (hunk 16 above)
Comment (liaarbel) [reply app/routers/todo_list.py]: fixed

File: app/templates/dayview.html (hunk 12 above)
Comment (liaarbel) [reply app/templates/dayview.html]: fixed

File: app/templates/dayview.html (hunk 11 above)
Comment (liaarbel) [reply app/templates/dayview.html]: fixed

File: app/routers/dayview.py [hunk 18]
```diff
@@ -212,6 +211,11 @@ async def dayview(
     )
     current_time_with_attrs = CurrentTimeAttributes(date=day)
     inter_day = international_days.get_international_day_per_day(session, day)
+    tasks = (session.query(Task)
+             .filter(Task.owner_id == user.user_id)
+             .filter(Task.date == day.date())
+             .order_by(Task.time)
+             )
```
Comment (yammesicka) [inline app/routers/dayview.py]: Please use precommit hooks to reformat

File: app/routers/todo_list.py [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+        db.commit()
+
+    except (SQLAlchemyError, TypeError):
+        return templates.TemplateResponse(
+            "calendar_day_view.html",
+            {"task_id": task_id},
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+        )
+    return RedirectResponse(
+        url=f"/day/{date_str}",
```
Comment (yammesicka) [inline app/routers/todo_list.py]: Use `url_for` on all those places

File: app/static/js/todo_list_modals.js [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+        entry.addEventListener('click', openEditModal);
+    });
+
+    document.getElementById('edit-modal-delete').addEventListener(
+        'click', deleteModal);
+});
+
+function openEditModal(event){
+    const modal = document.getElementById('edit-modal');
+    console.log(event);
```
Comment (yammesicka) [inline app/static/js/todo_list_modals.js]: Remove console.log parts

File: app/templates/partials/todo_list_modals/edit_task.html [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+        <div class="modal-dialog">
+            <div class="modal-content">
+                <div class="modal-header top-todo">
+                    <h1 id="edit-modal-label">
+                        Edit Task</h1>
+                    <button type="button" class="btn-close"
+                            data-bs-dismiss="modal"
+                            aria-label="Close"></button>
+                </div>
+                <form action="/task/edit" method="post">
```
Comment (yammesicka) [inline app/templates/partials/todo_list_modals/edit_task.html]: Use `url_for` wherever relevant

File: tests/test_todo_list.py [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+        owner_id=1,
+        owner=user,
+    )
+    yield task_example
+    delete_instance(session, task_example)
+
+
+def test_if_task_deleted(home_test_client, task1: Task, session):
+    response = home_test_client.post(
+        "/task/delete",
```
Comment (yammesicka) [inline tests/test_todo_list.py]: url_for

File: app/static/js/todo_list_modals.js [hunk 23]
```diff
@@ -0,0 +1,45 @@
+//function openEditModal(taskId, taskTime, taskTitle, taskDescription, taskImportant){
+//    $('#list-modal').modal('toggle');
+//    $('#edit-task-id').val(taskId);
+//    $('#customer-time2').val(taskTime);
+//    $('#customer-title2').val(taskTitle);
+//    $('#customer-descrip2').val(taskDescription);
+//    $('#is-important2').prop('checked', taskImportant);
+//    $('#edit-modal').modal('toggle');
+//}
```
Comment (yammesicka) [inline app/static/js/todo_list_modals.js]: Remove commented code

**Review by yammesicka — CHANGES_REQUESTED** [review]
Looks great :) Fix the minor issues and we're ready to go

File: app/templates/partials/calendar/calendar_base.html (hunk 17 above)
Comment (liaarbel) [reply app/templates/partials/calendar/calendar_base.html]: Weird... it's not appearing on my computer

File: app/routers/dayview.py (hunk 18 above)
Comment (liaarbel) [reply app/routers/dayview.py]: fixed

File: app/routers/todo_list.py (hunk 19 above)
Comment (liaarbel) [reply app/routers/todo_list.py]: fixed

File: app/static/js/todo_list_modals.js (hunk 20 above)
Comment (liaarbel) [reply app/static/js/todo_list_modals.js]: fixed

File: app/templates/partials/todo_list_modals/edit_task.html (hunk 21 above)
Comment (liaarbel) [reply app/templates/partials/todo_list_modals/edit_task.html]: fixed

File: tests/test_todo_list.py (hunk 22 above)
Comment (liaarbel) [reply tests/test_todo_list.py]: fixed

File: app/static/js/todo_list_modals.js (hunk 23 above)
Comment (liaarbel) [reply app/static/js/todo_list_modals.js]: fixed

**Review by yammesicka — APPROVED** [review]

**Pull request merged**